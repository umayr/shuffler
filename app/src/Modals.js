import { useEffect, useMemo, useState } from 'react'
import {
    Heading,
    List,
    Layer,
    Box,
    Button,
    RangeInput,
    Table,
    TableBody,
    TableCell,
    TableHeader,
    TableRow,
} from 'grommet'
import { Close as CloseIcon } from 'grommet-icons'

import { group, pick } from './service'

const Modal = ({ onClose, children }) => (
    <Layer onClickOutside={onClose} onEsc={onClose}>
        <Box
            pad={{
                horizontal: 'xlarge',
                vertical: 'large',
            }}
            gap="large"
            width={{ min: '50vw' }}
        >
            {children}
            <Button
                label="Close"
                icon={<CloseIcon />}
                secondary
                onClick={onClose}
            />
        </Box>
    </Layer>
)

const ACTION_TYPE_PICK = 'pick'
const ACTION_TYPE_GROUP = 'group'

const ACTION_TITLE = {
    [ACTION_TYPE_PICK]: 'Pick',
    [ACTION_TYPE_GROUP]: 'Group',
}

const ACTION_FUNC = {
    [ACTION_TYPE_PICK]: pick,
    [ACTION_TYPE_GROUP]: group,
}

const Action = ({ type, source, seed = '', onClose = () => {} }) => {
    if (type !== ACTION_TYPE_GROUP && type !== ACTION_TYPE_PICK) {
        throw new Error(`unsupported action type: ${type}`)
    }

    const [value, setValue] = useState(1)
    const [list, setList] = useState([])

    useEffect(() => {
        async function effect() {
            if (Array.isArray(source) && source.length === 0) {
                return
            }

            setList(
                await ACTION_FUNC[type](
                    source,
                    Number.parseInt(value, 10),
                    seed,
                ),
            )
        }

        effect()
    }, [source, value, type, seed])

    const data = useMemo(
        () => (type === ACTION_TYPE_GROUP ? list.map(v => v.join(',')) : list),
        [type, list],
    )

    return (
        <Modal onClose={onClose}>
            <Heading level={3} margin="xsmall">
                {ACTION_TITLE[type]} ({value})
            </Heading>
            <RangeInput
                value={value}
                max={source.length}
                min={1}
                onChange={event => setValue(event.target.value)}
            />
            {list.length > 0 && <List data={data} pad="small" />}
        </Modal>
    )
}

export const Pick = props => <Action type={ACTION_TYPE_PICK} {...props} />

export const Group = props => <Action type={ACTION_TYPE_GROUP} {...props} />

export const Help = ({ source, onClose = () => {} }) => (
    <Modal onClose={onClose}>
        <Heading level={3} margin="xsmall">
            Shortcuts
        </Heading>
        <Table>
            <TableHeader>
                <TableRow>
                    <TableCell scope="col" border="bottom">
                        Key
                    </TableCell>
                    <TableCell scope="col" border="bottom">
                        Action
                    </TableCell>
                </TableRow>
            </TableHeader>
            <TableBody>
                {source.map(({ key, text }) => (
                    <TableRow key={key}>
                        <TableCell scope="row">
                            <pre style={{ margin: 0 }}>{key}</pre>
                        </TableCell>
                        <TableCell>{text}</TableCell>
                    </TableRow>
                ))}
            </TableBody>
        </Table>
    </Modal>
)
