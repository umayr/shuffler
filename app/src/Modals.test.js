jest.mock('./service')

import { render, screen } from '@testing-library/react'

import { Help, Pick, Group } from './Modals'
import { Grommet } from 'grommet'

const setup = component =>
    render(component, {
        wrapper: ({ children }) => <Grommet plain>{children}</Grommet>,
    })

test('renders Help modal', () => {
    const { baseElement } = setup(
        <Help source={[{ key: '1', text: 'also known as one' }]} />,
    )

    expect(screen.getByText(/Shortcuts/)).toBeInTheDocument()
    expect(screen.getByText(/Key/)).toBeInTheDocument()
    expect(screen.getByText(/Action/)).toBeInTheDocument()
    expect(screen.getByText(/1/)).toBeInTheDocument()
    expect(screen.getByText(/also known as one/)).toBeInTheDocument()

    expect(baseElement).toMatchSnapshot()
})

test('renders Pick modal', () => {
    const { baseElement } = setup(<Pick source={[]} />)

    expect(screen.getByText(/Pick/)).toBeInTheDocument()

    expect(baseElement).toMatchSnapshot()
})

test('renders Group modal', () => {
    const { baseElement } = setup(<Group source={[]} />)

    expect(screen.getByText(/Group/)).toBeInTheDocument()

    expect(baseElement).toMatchSnapshot()
})
