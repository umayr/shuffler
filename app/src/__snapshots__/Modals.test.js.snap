// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Group modal 1`] = `
<body
  style="overflow: hidden;"
>
  <div />
  <div
    id="layerClone"
  >
    <div
      aria-hidden="false"
    >
      <div
        class="StyledLayer-rmtehz-0 kMZhng"
        tabindex="-1"
      >
        <div
          class="StyledLayer__StyledOverlay-rmtehz-1 buTrmF"
        />
        <div
          class="StyledLayer__StyledContainer-rmtehz-2 hPzbwk"
          data-g-portal-id="0"
          style="animation-direction: reverse;"
        >
          <a
            aria-hidden="true"
            class="LayerContainer__HiddenAnchor-sc-1srj14c-0 fxrbVi"
            tabindex="-1"
          />
          <div
            class="StyledBox-sc-13pk1d4-0 btaHDr"
          >
            <h3
              class="StyledHeading-sc-1rdh4aw-0 jCxtww"
            >
              Shortcuts
            </h3>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kVDkjZ"
            />
            <table
              class="StyledTable-sc-1m3u5g-6 hwKsrk"
            >
              <thead
                class="StyledTable__StyledTableHeader-sc-1m3u5g-4"
              >
                <tr
                  class="StyledTable__StyledTableRow-sc-1m3u5g-2"
                >
                  <th
                    class="StyledTable__StyledTableCell-sc-1m3u5g-0 HHgLC"
                    scope="col"
                  >
                    Key
                  </th>
                  <th
                    class="StyledTable__StyledTableCell-sc-1m3u5g-0 HHgLC"
                    scope="col"
                  >
                    Action
                  </th>
                </tr>
              </thead>
              <tbody
                class="StyledTable__StyledTableBody-sc-1m3u5g-3"
              >
                <tr
                  class="StyledTable__StyledTableRow-sc-1m3u5g-2"
                >
                  <th
                    class="StyledTable__StyledTableCell-sc-1m3u5g-0 hkXmPk"
                    scope="row"
                  >
                    <pre
                      style="margin: 0px;"
                    >
                      1
                    </pre>
                  </th>
                  <td
                    class="StyledTable__StyledTableCell-sc-1m3u5g-0 hkXmPk"
                  >
                    also known as one
                  </td>
                </tr>
              </tbody>
            </table>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kVDkjZ"
            />
            <button
              class="StyledButton-sc-323bzc-0 kflPHx"
              type="button"
            >
              <div
                class="StyledBox-sc-13pk1d4-0 jncmfT"
              >
                <svg
                  aria-label="Close"
                  class="StyledIcon-ofa7kd-0 hVBrNB"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m3 3 18 18M3 21 21 3"
                    fill="none"
                    stroke="#000"
                    stroke-width="2"
                  />
                </svg>
                <div
                  class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kaiRIg"
                />
                Close
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div />
  <div
    id="layerClone"
  >
    <div
      aria-hidden="false"
    >
      <div
        class="StyledLayer-rmtehz-0 kMZhng"
        tabindex="-1"
      >
        <div
          class="StyledLayer__StyledOverlay-rmtehz-1 buTrmF"
        />
        <div
          class="StyledLayer__StyledContainer-rmtehz-2 hPzbwk"
          data-g-portal-id="0"
          style="animation-direction: reverse;"
        >
          <a
            aria-hidden="true"
            class="LayerContainer__HiddenAnchor-sc-1srj14c-0 fxrbVi"
            tabindex="-1"
          />
          <div
            class="StyledBox-sc-13pk1d4-0 btaHDr"
          >
            <h3
              class="StyledHeading-sc-1rdh4aw-0 jCxtww"
            >
              Pick
               (
              1
              )
            </h3>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kVDkjZ"
            />
            <input
              class="StyledRangeInput-sc-15st9ck-0 keldXR"
              max="0"
              min="1"
              type="range"
              value="1"
            />
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kVDkjZ"
            />
            <button
              class="StyledButton-sc-323bzc-0 kflPHx"
              type="button"
            >
              <div
                class="StyledBox-sc-13pk1d4-0 jncmfT"
              >
                <svg
                  aria-label="Close"
                  class="StyledIcon-ofa7kd-0 hVBrNB"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m3 3 18 18M3 21 21 3"
                    fill="none"
                    stroke="#000"
                    stroke-width="2"
                  />
                </svg>
                <div
                  class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kaiRIg"
                />
                Close
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      class="StyledGrommet-sc-19lkkz7-0 cXIcsV"
    />
  </div>
  <div>
    <div
      aria-hidden="false"
    >
      <div
        class="StyledLayer-rmtehz-0 kMZhng"
        tabindex="-1"
      >
        <div
          class="StyledLayer__StyledOverlay-rmtehz-1 buTrmF"
        />
        <div
          class="StyledLayer__StyledContainer-rmtehz-2 hPzbwk"
          data-g-portal-id="0"
        >
          <a
            aria-hidden="true"
            class="LayerContainer__HiddenAnchor-sc-1srj14c-0 fxrbVi"
            tabindex="-1"
          />
          <div
            class="StyledBox-sc-13pk1d4-0 btaHDr"
          >
            <h3
              class="StyledHeading-sc-1rdh4aw-0 jCxtww"
            >
              Group
               (
              1
              )
            </h3>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kVDkjZ"
            />
            <input
              class="StyledRangeInput-sc-15st9ck-0 keldXR"
              max="0"
              min="1"
              type="range"
              value="1"
            />
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kVDkjZ"
            />
            <button
              class="StyledButton-sc-323bzc-0 kflPHx"
              type="button"
            >
              <div
                class="StyledBox-sc-13pk1d4-0 jncmfT"
              >
                <svg
                  aria-label="Close"
                  class="StyledIcon-ofa7kd-0 hVBrNB"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m3 3 18 18M3 21 21 3"
                    fill="none"
                    stroke="#000"
                    stroke-width="2"
                  />
                </svg>
                <div
                  class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kaiRIg"
                />
                Close
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`renders Help modal 1`] = `
<body
  style="overflow: hidden;"
>
  <div>
    <div
      class="StyledGrommet-sc-19lkkz7-0 cXIcsV"
    />
  </div>
  <div>
    <div
      aria-hidden="false"
    >
      <div
        class="StyledLayer-rmtehz-0 kMZhng"
        tabindex="-1"
      >
        <div
          class="StyledLayer__StyledOverlay-rmtehz-1 buTrmF"
        />
        <div
          class="StyledLayer__StyledContainer-rmtehz-2 hPzbwk"
          data-g-portal-id="0"
        >
          <a
            aria-hidden="true"
            class="LayerContainer__HiddenAnchor-sc-1srj14c-0 fxrbVi"
            tabindex="-1"
          />
          <div
            class="StyledBox-sc-13pk1d4-0 btaHDr"
          >
            <h3
              class="StyledHeading-sc-1rdh4aw-0 jCxtww"
            >
              Shortcuts
            </h3>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kVDkjZ"
            />
            <table
              class="StyledTable-sc-1m3u5g-6 hwKsrk"
            >
              <thead
                class="StyledTable__StyledTableHeader-sc-1m3u5g-4"
              >
                <tr
                  class="StyledTable__StyledTableRow-sc-1m3u5g-2"
                >
                  <th
                    class="StyledTable__StyledTableCell-sc-1m3u5g-0 HHgLC"
                    scope="col"
                  >
                    Key
                  </th>
                  <th
                    class="StyledTable__StyledTableCell-sc-1m3u5g-0 HHgLC"
                    scope="col"
                  >
                    Action
                  </th>
                </tr>
              </thead>
              <tbody
                class="StyledTable__StyledTableBody-sc-1m3u5g-3"
              >
                <tr
                  class="StyledTable__StyledTableRow-sc-1m3u5g-2"
                >
                  <th
                    class="StyledTable__StyledTableCell-sc-1m3u5g-0 hkXmPk"
                    scope="row"
                  >
                    <pre
                      style="margin: 0px;"
                    >
                      1
                    </pre>
                  </th>
                  <td
                    class="StyledTable__StyledTableCell-sc-1m3u5g-0 hkXmPk"
                  >
                    also known as one
                  </td>
                </tr>
              </tbody>
            </table>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kVDkjZ"
            />
            <button
              class="StyledButton-sc-323bzc-0 kflPHx"
              type="button"
            >
              <div
                class="StyledBox-sc-13pk1d4-0 jncmfT"
              >
                <svg
                  aria-label="Close"
                  class="StyledIcon-ofa7kd-0 hVBrNB"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m3 3 18 18M3 21 21 3"
                    fill="none"
                    stroke="#000"
                    stroke-width="2"
                  />
                </svg>
                <div
                  class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kaiRIg"
                />
                Close
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`renders Pick modal 1`] = `
<body
  style="overflow: hidden;"
>
  <div />
  <div
    id="layerClone"
  >
    <div
      aria-hidden="false"
    >
      <div
        class="StyledLayer-rmtehz-0 kMZhng"
        tabindex="-1"
      >
        <div
          class="StyledLayer__StyledOverlay-rmtehz-1 buTrmF"
        />
        <div
          class="StyledLayer__StyledContainer-rmtehz-2 hPzbwk"
          data-g-portal-id="0"
          style="animation-direction: reverse;"
        >
          <a
            aria-hidden="true"
            class="LayerContainer__HiddenAnchor-sc-1srj14c-0 fxrbVi"
            tabindex="-1"
          />
          <div
            class="StyledBox-sc-13pk1d4-0 btaHDr"
          >
            <h3
              class="StyledHeading-sc-1rdh4aw-0 jCxtww"
            >
              Shortcuts
            </h3>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kVDkjZ"
            />
            <table
              class="StyledTable-sc-1m3u5g-6 hwKsrk"
            >
              <thead
                class="StyledTable__StyledTableHeader-sc-1m3u5g-4"
              >
                <tr
                  class="StyledTable__StyledTableRow-sc-1m3u5g-2"
                >
                  <th
                    class="StyledTable__StyledTableCell-sc-1m3u5g-0 HHgLC"
                    scope="col"
                  >
                    Key
                  </th>
                  <th
                    class="StyledTable__StyledTableCell-sc-1m3u5g-0 HHgLC"
                    scope="col"
                  >
                    Action
                  </th>
                </tr>
              </thead>
              <tbody
                class="StyledTable__StyledTableBody-sc-1m3u5g-3"
              >
                <tr
                  class="StyledTable__StyledTableRow-sc-1m3u5g-2"
                >
                  <th
                    class="StyledTable__StyledTableCell-sc-1m3u5g-0 hkXmPk"
                    scope="row"
                  >
                    <pre
                      style="margin: 0px;"
                    >
                      1
                    </pre>
                  </th>
                  <td
                    class="StyledTable__StyledTableCell-sc-1m3u5g-0 hkXmPk"
                  >
                    also known as one
                  </td>
                </tr>
              </tbody>
            </table>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kVDkjZ"
            />
            <button
              class="StyledButton-sc-323bzc-0 kflPHx"
              type="button"
            >
              <div
                class="StyledBox-sc-13pk1d4-0 jncmfT"
              >
                <svg
                  aria-label="Close"
                  class="StyledIcon-ofa7kd-0 hVBrNB"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m3 3 18 18M3 21 21 3"
                    fill="none"
                    stroke="#000"
                    stroke-width="2"
                  />
                </svg>
                <div
                  class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kaiRIg"
                />
                Close
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
  <div>
    <div
      class="StyledGrommet-sc-19lkkz7-0 cXIcsV"
    />
  </div>
  <div>
    <div
      aria-hidden="false"
    >
      <div
        class="StyledLayer-rmtehz-0 kMZhng"
        tabindex="-1"
      >
        <div
          class="StyledLayer__StyledOverlay-rmtehz-1 buTrmF"
        />
        <div
          class="StyledLayer__StyledContainer-rmtehz-2 hPzbwk"
          data-g-portal-id="0"
        >
          <a
            aria-hidden="true"
            class="LayerContainer__HiddenAnchor-sc-1srj14c-0 fxrbVi"
            tabindex="-1"
          />
          <div
            class="StyledBox-sc-13pk1d4-0 btaHDr"
          >
            <h3
              class="StyledHeading-sc-1rdh4aw-0 jCxtww"
            >
              Pick
               (
              1
              )
            </h3>
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kVDkjZ"
            />
            <input
              class="StyledRangeInput-sc-15st9ck-0 keldXR"
              max="0"
              min="1"
              type="range"
              value="1"
            />
            <div
              class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kVDkjZ"
            />
            <button
              class="StyledButton-sc-323bzc-0 kflPHx"
              type="button"
            >
              <div
                class="StyledBox-sc-13pk1d4-0 jncmfT"
              >
                <svg
                  aria-label="Close"
                  class="StyledIcon-ofa7kd-0 hVBrNB"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="m3 3 18 18M3 21 21 3"
                    fill="none"
                    stroke="#000"
                    stroke-width="2"
                  />
                </svg>
                <div
                  class="StyledBox__StyledBoxGap-sc-13pk1d4-1 kaiRIg"
                />
                Close
              </div>
            </button>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
